libyara/grammar.y

```diff
@@ -601,40 +604,89 @@ string_declaration
 
 
 string_modifiers
-    : /* empty */                         { $$ = 0; }
-    | string_modifiers string_modifier    { $$ = $1 | $2; }
+    : /* empty */
+      {
+        $$.flags = 0;
+        $$.xor_min = 0;
+        $$.xor_max = 0;
+      }
+    | string_modifiers string_modifier
+      {
+        $$.flags = $1.flags | $2.flags;
```

Comment:
Here you are initializing `$$.xor_min` and `$$.xor_max` only when `$2.flags` has the `STRING_GFLAGS_XOR` flag, but what if `$1.flags` is the one that has `STRING_GFLAGS_XOR`? The values for `xor_min` and `xor_max` are not copied to the result in that case. The interesting thing is that there is a test case that should cover this and seems to be working fine, so I may be missing something.

Also, this leads me to another problem that we didn't had until now. What if the `xor` modifier is written twice with different arguments? Right now YARA does't prevents the duplication of a modifier, you can write:

```
  $a = "foo" wide wide 
```

That wasn't a problem until now because modifiers were just flags, but now that they can accept arguments the semantics are more complex. As this is a very unusual case, because it doesn't make sense to write the same modifier twice, I think we can keep the current semantics of accepting repeated modifiers and taking into account only the last occurrence. Just wanted to bring attention to this.

Reply:
I found why this is working. The answer is here:
https://github.com/wxsBSD/yara/blob/xor_fixes_2/libyara/grammar.c#L1689

The default action for a reduction rule when you don't provide one is `$$ = $1`, but interestingly enough bison does this in *all cases* before executing the user provided action. So, you are getting `$$` already initialized with the value of `$1` and this is why it works. However, stated in the bison's comment, this is not documented and we shouldn't rely on this behavior. We most assume that `$$` is always uninitialized .

Reply:
> Here you are initializing $$.xor_min and $$.xor_max only when $2.flags has the STRING_GFLAGS_XOR flag, but what if $1.flags is the one that has STRING_GFLAGS_XOR? The values for xor_min and xor_max are not copied to the result in that case. The interesting thing is that there is a test case that should cover this and seems to be working fine, so I may be missing something.

My understanding was `$1` is the accumulation of all the other modifiers, so if it has `STRING_GFLAGS_XOR` flag we would have already processed it.

> Also, this leads me to another problem that we didn't had until now. What if the xor modifier is written twice with different arguments? 

I actually think it should be an error but wanted to leave it up to you. IMO, we should reduce ambiguity in the language. I'll update it to make this an error and add a test case for it too.

I'll take another pass through this code tonight and clean it up so that:

1. There are no changes to the lexer.
2. Duplicate modifiers are errors.
3. We don't rely on the undocumented automatic reduction of rules.

Reply:
> > Here you are initializing $$.xor_min and $$.xor_max only when $2.flags has the STRING_GFLAGS_XOR flag, but what if $1.flags is the one that has STRING_GFLAGS_XOR? The values for xor_min and xor_max are not copied to the result in that case. The interesting thing is that there is a test case that should cover this and seems to be working fine, so I may be missing something.
> 
> My understanding was `$1` is the accumulation of all the other modifiers, so if it has `STRING_GFLAGS_XOR` flag we would have already processed it.
> 

Yes, that's correct, `$1` is the accumulation of all the other modifiers. But `$1` could have the XOR flag set and some values in xor_min and xor_max, but if `$2` doesn't have the flag, the values already `$1` are not copied into `$$`. The values in `$$` are the result from this rule, and could be undefined if not explicitly set.

It works because the implementation detail in Bison that forces `$$ = $1` before calling our code. But this is not documented, in theory if you don't set the result value for your reduction rule in `$$` it is undefined.

> > Also, this leads me to another problem that we didn't had until now. What if the xor modifier is written twice with different arguments?
> 
> I actually think it should be an error but wanted to leave it up to you. IMO, we should reduce ambiguity in the language. I'll update it to make this an error and add a test case for it too.
> 
> I'll take another pass through this code tonight and clean it up so that:
> 
> 1. There are no changes to the lexer.
> 2. Duplicate modifiers are errors.
> 3. We don't rely on the undocumented automatic reduction of rules.

Ok, a I agree on making it an error. I hope there are no rules with duplicated modifiers out there.